### ISSUE TITLE
Spawning with JupyterHub/Kubespawner

### ISSUE THREAD

**Issue by michzimny** [sciencemesh/cs3api4lab#9]
My attempts were done against https://github.com/JarCz/cs3api4lab/tree/docker

I've been trying to run that jupyterlab image with JupyterHub 1.1 and KubeSpawner 0.11 (which are available within JupyterHub helm chart v. 0.9.0).

# My attempts

Here go my attempts. My JupyterHub setup, for reproducibility, is given at the end of this description.

## 1. Original Dockerfile without command

The original Dockerfile (at [that commit](https://github.com/JarCz/cs3api4lab/commit/5beedd64871e1dc71038a32b07aa3b4329f9050b)) overrides `ENTRYPOINT`. So I tried to run it with empty (null) command injected by the spawner. It ends up with [these logs](https://pastebin.com/65G5MpnW) and "404 : Not Found" from the Hub.

I think this is because in such a setup, the hub and the notebook container can't communicate regarding the notebook's authz token (that is only printed in the notebook's logs).

Moreover, I think overriding the `ENTRYPOINT` in our Dockerfile is risky. It is also not so compatible with [the base image](https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile) that uses `tini` as entrypoint, and the actual jupyter command goes in `CMD`. I think we should do the same (if we need to override the command at all, see "Side note" below).

## 2. Original Dockerfile and injecting command

If I allow spawner to inject its default command to this container (which is either `jupyterhub-singleuser` or `jupyter-labhub`, tested both), starting the container ends up with:

```
[C 07:33:41.332 LabApp] No such file or directory: /home/jovyan/jupyterhub-singleuser
```

I guess this is because the injected command is added as a parameter of the `ENTRYPOINT`. So again, we should not have this `ENTRYPOINT`.

## 3. ENTRYPOINT changed to CMD

For this test, I made that change - https://github.com/michzimny/cs3api4lab/commit/a6164df47bc6f783f2bfc58a9a80641d6604af67 - as justified above.

If I spawn it without injecting any command (i.e. the Dockerfile's `CMD` is called), it ends up as in point 1 above.

If I spawn it with command set to `jupyterhub-singleuser`, then it starts up properly, but the plugin seems not to be enabled (see [logs here](https://pastebin.com/7VwV6MpK)). I expected to see any errors related to connecting to non-existing IOP (as I can see when running locally). In the file tree, I simply see local files. However, when clicking Extension Manager, I see the plugin enabled.

We should probably investigate why the plugin is not used in this setup.

# Side note on startup command

Do we need to override `ENTRYPOINT`/`CMD` at all? Running `/usr/local/bin/start-notebook.sh` (see [the base image](https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile)) is the default container command for the jupyter/docker-stacks images, and it works fine for containers run locally as well as for run by JupyterHub. I think we should not override it unless we have a good reason.

# My environment

(Based on [Z2JH](https://zero-to-jupyterhub.readthedocs.io/en/latest/setup-jupyterhub/setup-jupyterhub.html))

- Have K8S cluster. 
- Install Helm v3. locally
- Prepare `hub.yaml`:

```
proxy:
  secretToken: ...
  https:
    enabled: false

singleuser:
  image:
    name: michzimny/cs3api4lab-jupyterlab
    tag: branch-docker  # differs in different attempts
    pullPolicy: Always
  # cmd: null  # differs in different attempts, if commented out, it falls back to default "jupyterhub-singleuser"
  defaultUrl: "/lab"
```

- Then: 

```
helm upgrade -i jhub jupyterhub/jupyterhub --namespace $NS --version=0.9.0 -f hub.yaml --cleanup-on-fail
```

**Comment by michzimny** [comment]
CC @JarCz @diocas

**Comment by diocas** [comment]
I agree with Michal: we should not change `ENTRYPOINT`.
Just install what you have to install, fix permissions, return to `$HOME`, switch to `jovyan` and don't do anything else. If this is done like that, then this image should be usable in any deployment.

**Comment by diocas** [comment]
Btw Michal, `jupyter-labhub` should be deprecated by now, I think.

**Comment by JarCz** [comment]
@michzimny I have updated Dockerfile. I deleted entrypoint and enabled Lab at default in environment value, because "notebook" don't load cs3api4lab extension.

**Comment by JarCz** [comment]
@michzimny I tested my version of docker image with steps:

1. Have a minikube and Helm 3

2. Create docker image based on commit: https://github.com/JarCz/cs3api4lab/commit/9e542c8920caf04b57fb815728cf77d4bca5fcfc 

3. Prepare hub.yam file:
```
proxy:
  secretToken: "…"
  https:
    enabled: false

singleuser:
  image:
    name: jarczsm/cs3api4lab
    tag: test  # differs in different attempts
    pullPolicy: Always
  # cmd: null  # differs in different attempts, if commented out, it falls back to default "jupyterhub-singleuser"
  defaultUrl: "/lab"
```
3. Add jupyterhub repo to helm:
```
helm repo add jupyterhub https://jupyterhub.github.io/helm-chart/
helm repo update
```

4. Install container jupyterhub and lab with extension:

```
helm upgrade -i jhub jupyterhub/jupyterhub --namespace jhub --create-namespace --version=0.9.0 -f hub.yaml –cleanup-on-fail
```

At shell pod, I'm see the extension is properly load:
```
jovyan@jupyter-test:/opt/conda$ jupyter serverextension list
config dir: /opt/conda/etc/jupyter
    jupyterlab  enabled 
    - Validating...
      jupyterlab 2.2.7 OK
    cs3api4lab  enabled 
    - Validating...
      cs3api4lab 0.1.0 OK
```
Jupyter with Lab can open in browser

**Comment by michzimny** [comment]
Thanks for the update. I will be able to try it early next week.

> Btw Michal, jupyter-labhub should be deprecated by now, I think.

Ok, right.

**Comment by michzimny** [comment]
@JarCz, I made a test of the new image. The extension is properly loaded, as in your case.

However, I did not provide any configuration (like IOP connection details). What is the expected outcome in my case? In the file manager, I see only local files. In logs, I get the following warning:

```
[W ... SingleUserNotebookApp commands:2229] Failed to fetch package metadata for '@sciencemesh/cs3api4lab': <HTTPError 404: 'Not Found'>
```

What should I do next? A proper test with an existing IOP?

Can I use any publicly available IOP instance in my test? How to? CC @diocas 

Or what is the easiest way to set up an own instance?
